### DISCUSSION TITLE
[Feature] Split Person just as we have merge Persons for face recognition

### DISCUSSION THREAD

**Discussion by aniruddhas** [immich-app/immich#7672] [Feature Request]
### The feature

[I don't know if I'm facing a misconfiguration or bug or a unique scenario. Hence put in discussions. ]
Thanks a lot for this project,  Immich Team! Logged in github after a long time to say thanks /star. Then recalled this issue I'm facing. Please,  be kind if don't provide all required information immediately.  

Background:
I've been on immich for personal photos since about last 5/6 months.
Completely switched to immich other than the immediate competitors (had given up on who couldn't be named quite sometime back way before that) . Couldn't stay with the immediate competitor for multiple reasons. 

Throughout my immich journey, with the will to stay here,  I faced lots of issues. Most of them got resolved because of active development and community responses. None were a big deal.

I wanted to file bugs, they were already there and the next day / week (at most) poof!!!!

Issue at hand:
Had to re import all photos because of old enclosures died on me.
My kids (4/5 years age difference ) are getting matched as the same person. 
In previous versions,  I used to find single different picture (tagged as a different person) and merge with that. Photo by photo or batches. 

As of now, all photos of the kids are marked as the same person.  Many photos have both of them in the frame and still tagged as same person. 

How do I deal with this? 
Should I change the sensitivity of matches or something?

Is anyone else facing this?

Thanks a lot again! 

Apologies, if thus isn't the expected format.

I'm on 1.97.0 on server. 
On app it's 1.97.0 build 125.

### Platform

- [X] Server
- [ ] Web
- [ ] Mobile

### Edit: Additional information
- FACIAL RECOGNITION MODEL    :  buffalo_l
- MIN DETECTION SCORE               : 0.7
- MAX RECOGNITION DISTANCE    : 0.6
- MIN RECOGNIZED FACES             : 0.3

**Comment by arhills** [comment]
I too have had this issue, with pictures of my brother and myself as children (55 years ago :) ). I also had this issue with my brothers in law who are identical twins, even if they are photographed together.

**Reply by aniruddhas** [reply]
Oh. In my case, both of kids (son and daughter) are mapped as single person. Whether there are separate photographs or single frame containing both of them.

**Reply by arhills** [reply]
Yep I have the same, just thought it stranger if they were in the same photo.

**Reply by aniruddhas** [reply]
Might have to play with following settings, if someone learned from the community / immich team responds in favor of that:

```

MIN DETECTION SCORE : 0.7
MAX RECOGNITION DISTANCE : 0.6
```
But, having face recognition run on thousands of photos & videos is the part I'm getting jittery about. This current import took me almost a week on a decent laptop.

**Reply by aviv926** [reply]
You can try the offer that appears here https://github.com/immich-app/immich/discussions/3946

Starting with the next release it will automatically operate on a max recognition distance of 0.5 #7689

**Reply by aniruddhas** [reply]
Thanks a lot. Will try to reduce distance to 0.5 , rerun the recognition and then report back here.
Having to run the rescans on entire library is gonna take some time though. Wish, there was a way to just run on specific assets where concerned faces were identified.

**Reply by bo0tzz** [reply]
When changing the distance you shouldn't need to rerun detection, only recognition, which should be pretty fast.

**Reply by aniruddhas** [reply]
Arghh. Had triggered detection before I saw your reply.  I would keep this mind! 
I'm not sure, if I read this in any document.  
Can you point me somewhere?

**Reply by aviv926** [reply]
> Arghh. Had triggered detection before I saw your reply.  I would keep this mind! 
> I'm not sure, if I read this in any document.  
> Can you point me somewhere? 

This is going to appear in the documentation soon.
https://github.com/immich-app/immich/pull/7500

**Reply by aniruddhas** [reply]
Awesome!

**Reply by aviv926** [reply]
> Awesome!

Can you update if the issue is resolved so we can mark it as resolved?

**Reply by aniruddhas** [reply]
Will update once resolved. 
Face detection phase is over. 
It is currently stuck at Face Recognition:
```
Active 1
Waiting  1,31,835
```

Pausing and resuming didn't do a dent to the Waiting number. Will report back in a few hours again.

**Reply by aniruddhas** [reply]
Tried to clear the "Face Recognition Queue" 
No people seen in "People" tab.
Tried  to detect missing in "Face Detection" in order to queue the faces. It didn't do anything.

Face recognition control says:
```
 FACIAL RECOGNITION

Group detected faces into people. This step runs after Face Detection is complete. "All" (re-)clusters all faces. "Missing" queues faces that don't have a person assigned.
```

**Reply by aviv926** [reply]
> Tried to clear the "Face Recognition Queue" 
> No people seen in "People" tab.
> Tried  to detect missing in "Face Detection" in order to queue the faces. It didn't do anything.
> 
> Face recognition control says:
> ```
>  FACIAL RECOGNITION
> 
> Group detected faces into people. This step runs after Face Detection is complete. "All" (re-)clusters all faces. "Missing" queues faces that don't have a person assigned.
> ```
> 

It's fine as far as I know just give it enough time to run. If you change the log level in the admin settings you can see that it works to find identical faces from the embeds until everything is finished.

**Reply by aniruddhas** [reply]
Anyway, I had hit the redo button. Will have to wait for at least completion. Will report back after that.

**Reply by aniruddhas** [reply]
Finally, I could nudge it to move. But, issue persists with 0.5 distance as well. I will wait for completion of entire Facial recognition.
As of now, both kids are tagged as same.

Probably, it was stuck   because of extremely large collection and low processing power.
Details:
I can see in logs, microservices is reporting that it is processing correctly now:
```
{"log":"\u001b[32m[Nest] 7  - \u001b[39m03/09/2024, 4:17:01 AM \u001b[32m    LOG\u001b[39m \u001b[38;5;3m[PersonService] \u001b[39m\u001b[32mCreating new person for face 034105d6-48d7-4557-b2b9-ab52d77c9bf7\u001b[39m\n","stream":"stdout","time":"2024-03-09T04:17:01.867219665Z"}
{"log":"\u001b[32m[Nest] 7  - \u001b[39m03/09/2024, 4:17:02 AM \u001b[32m    LOG\u001b[39m \u001b[38;5;3m[PersonService] \u001b[39m\u001b[32mCreating new person for face b5e41dbc-bd97-4a13-9477-169b3cd32c3c\u001b[39m\n","stream":"stdout","time":"2024-03-09T04:17:02.264075815Z"}
{"log":"\u001b[32m[Nest] 7  - \u001b[39m03/09/2024, 4:17:02 AM \u001b[32m    LOG\u001b[39m \u001b[38;5;3m[PersonService] \u001b[39m\u001b[32mCreating new person for face dd87cd73-9f52-4135-9e55-54efe3897386\u001b[39m\n","stream":"stdout","time":"2024-03-09T04:17:02.587200356Z"}
{"log":"\u001b[32m[Nest] 7  - \u001b[39m03/09/2024, 4:17:09 AM \u001b[32m    LOG\u001b[39m \u001b[38;5;3m[PersonService] \u001b[39m\u001b[32mCreating new person for face 9a0d1eda-c216-41c5-8815-4c620707bfac\u001b[39m\n","stream":"stdout","time":"2024-03-09T04:17:09.062069581Z"}
{"log":"\u001b[32m[Nest] 7  - \u001b[39m03/09/2024, 4:17:09 AM \u001b[32m    LOG\u001b[39m \u001b[38;5;3m[PersonService] \u001b[39m\u001b[32mCreating new person for face 77cc19ae-92a7-4ec1-824b-7fc4cebbde58\u001b[39m\n","stream":"stdout","time":"2024-03-09T04:17:09.621448212Z"}
{"log":"\u001b[32m[Nest] 7  - \u001b[39m03/09/2024, 4:17:10 AM \u001b[32m    LOG\u001b[39m \u001b[38;5;3m[PersonService] \u001b[39m\u001b[32mCreating new person for face 450182d6-fc5d-465e-b170-b9dcd5f1642c\u001b[39m\n","stream":"stdout","time":"2024-03-09T04:17:10.182262792Z"}
{"log":"\u001b[32m[Nest] 7  - \u001b[39m03/09/2024, 4:17:10 AM \u001b[32m    LOG\u001b[39m \u001b[38;5;3m[PersonService] \u001b[39m\u001b[32mCreating new person for face b9d90a8f-71ba-44d4-98cd-10bc1ddb060f\u001b[39m\n","stream":"stdout","time":"2024-03-09T04:17:10.565189826Z"}

```

Earlier microservices was showing these:

```
[Nest] 7  - 03/03/2024, 1:55:43 PM   ERROR [JobService] Object:
{
  "id": "8797c207-c9f3-4c7d-9048-03becf204f29"
}

[Nest] 7  - 03/03/2024, 1:55:43 PM   ERROR [JobService] Unable to run job handler (migration/migrate-asset): Error: timeout exceeded when trying to connect
[Nest] 7  - 03/03/2024, 1:55:43 PM   ERROR [JobService] Error: timeout exceeded when trying to connect
    at Timeout._onTimeout (/usr/src/app/node_modules/pg-pool/index.js:205:27)
    at listOnTimeout (node:internal/timers:573:17)
    at process.processTimers (node:internal/timers:514:7)
[Nest] 7  - 03/03/2024, 1:55:43 PM   ERROR [JobService] Object:
{
  "id": "3c1d6a7d-d19c-4140-b675-e8486323d39f"

```
And immich_ml had these:
```

[03/08/24 18:33:47] ERROR    Exception in ASGI application                      
                                                                                
                             ╭─────── Traceback (most recent call last) ───────╮
                             │ /usr/src/app/main.py:108 in predict             │
                             │                                                 │
                             │   105 │                                         │
                             │   106 │   model = await load(await model_cache. │
                             │   107 │   model.configure(**kwargs)             │
                             │ ❱ 108 │   outputs = await run(model.predict, in │
                             │   109 │   return ORJSONResponse(outputs)        │
                             │   110                                           │
                             │   111                                           │
                             │                                                 │
                             │ /usr/src/app/main.py:115 in run                 │
                             │                                                 │
                             │   112 async def run(func: Callable[..., Any], i │
                             │   113 │   if thread_pool is None:               │
                             │   114 │   │   return func(inputs)               │
                             │ ❱ 115 │   return await asyncio.get_running_loop │
                             │   116                                           │
                             │   117                                           │
                             │   118 async def load(model: InferenceModel) ->  │
                             │                                                 │
                             │ /usr/local/lib/python3.11/concurrent/futures/th │
                             │ read.py:58 in run                               │
                             │                                                 │
                             │ /usr/src/app/models/base.py:60 in predict       │
                             │                                                 │
                             │    57 │   │   self.load()                       │
                             │    58 │   │   if model_kwargs:                  │
                             │    59 │   │   │   self.configure(**model_kwargs │
                             │ ❱  60 │   │   return self._predict(inputs)      │
                             │    61 │                                         │
                             │    62 │   @abstractmethod                       │
                             │    63 │   def _predict(self, inputs: Any) -> An │
                             │                                                 │
                             │ /usr/src/app/models/facial_recognition.py:45 in │
                             │ _predict                                        │
                             │                                                 │
                             │   42 │                                          │
                             │   43 │   def _predict(self, image: NDArray[np.u │
                             │   44 │   │   if isinstance(image, bytes):       │
                             │ ❱ 45 │   │   │   decoded_image = cv2.imdecode(n │
                             │      cv2.IMREAD_COLOR)                          │
                             │   46 │   │   else:                              │
                             │   47 │   │   │   decoded_image = image          │
                             │   48 │   │   assert is_ndarray(decoded_image, n │
                             ╰─────────────────────────────────────────────────╯
                             error: OpenCV(4.9.0)                               
                             /io/opencv/modules/imgcodecs/src/loadsave.cpp:801: 
                             error: (-215:Assertion failed) !buf.empty() in     
                             function 'imdecode_'                               
                                                                                
[03/08/24 22:29:58] INFO     Shutting down due to inactivity.                   
[03/08/24 22:29:58] INFO     Shutting down                                      
[03/08/24 22:29:58] INFO     Waiting for application shutdown.                  
[03/08/24 22:29:59] INFO     Application shutdown complete.                     
[03/08/24 22:29:59] INFO     Finished server process [1486]                     
[03/08/24 22:29:59] INFO     Worker exiting (pid: 1486)                         
[03/08/24 22:29:59] INFO     Booting worker with pid: 2494                      
[03/08/24 22:30:05] INFO     Started server process [2494]                      
[03/08/24 22:30:05] INFO     Waiting for application startup.                   
[03/08/24 22:30:05] INFO     Created in-memory cache with unloading after 300s  
                             of inactivity.                                     
[03/08/24 22:30:05] INFO     Initialized request thread pool with 4 threads.    
[03/08/24 22:30:05] INFO     Application startup complete.                      
```
I killed immich_microservices as an attempt to get it out of the retry loop. Immich ML was waiting for jobs and immich microservices wasn't able to reach ML for whatever reason. As soon as it was killed, it got auto restarted and the waiting numbers in Facial recognition started going down from 1,31,835 or something.
Uptimes for reference (5 days ago, had upgraded to 1.97.0)
```
76abe247ef81   ghcr.io/immich-app/immich-server:release             "tini -- /bin/bash s…"   5 days ago     Up 5 days               0.0.0.0:2283->3001/tcp   immich_server
7ca1adec8390   ghcr.io/immich-app/immich-server:release             "tini -- /bin/bash s…"   5 days ago     Up 14 minutes           3001/tcp                 immich_microservices
c7e236d5820b   tensorchord/pgvecto-rs:pg14-v0.2.0                   "docker-entrypoint.s…"   5 days ago     Up 5 days               5432/tcp                 immich_postgres
b2d86140899c   ghcr.io/immich-app/immich-machine-learning:release   "tini -- ./start.sh"     5 days ago     Up 5 days                                        immich_machine_learning
2db2e1d170d2   redis:6.2-alpine                                     "docker-entrypoint.s…"   5 days ago     Up 5 days               6379/tcp                 immich_redis

```

**Reply by aniruddhas** [reply]
Came back to report, reducing distance even further helped.
In previous runs, with distance of 0.5 ML still assumed both kids to be same.
With following relevant settings, everything is working out as expected:
```
    FACIAL RECOGNITION MODEL : buffalo_l
    MIN DETECTION SCORE : 0.7
    MAX RECOGNITION DISTANCE : 0.4
```
Thanks for the help! Kindly, close this discussion.

**Comment by aniruddhas** [comment]
With max recognition distance of 0.4 , now I don't get two different people merged as the same person. At the same time without the ability to split the person  (doesn't make sense to build this feature as it may turn too complex too soon, considering UI, backend and batching ability it may require. ). I am happy with one real person tagged as two people and have correct assets with those two or more groups. I have the ability to merge them. 
Apologies to everyone,  as this dragged for quite some time. Had to retest / retry quite a few things as described in this discussion against the large set. 
Thanks again,  Immich Team! 
Closing this discussion.

**Comment by ryanjdillon** [comment]
I haven't yet looked to see what the data model looks like for a person, but I think adding a foreign ref to merged persons could allow for this feature and potentially allow more robust subsequent identification of photos of a person from different life stages.

That way after merge the merged persons still exist as DB items, they get turned to hide (or perhaps a new visibility setting exists based on a populated `merge_ids` field. Then the model has a better target for child vs. adult (I would think), and then the merge could be undone.

For example from this:
```
Person
* id
* name
* model target
* hidden
```
To this:
```
Person
* id
* merge_ids
* name
* model target
* hidden
```

This wouldn't cause any ugly DB migrations, but folks who've already merged the old way couldn't undo past merges of course without a re-scan and the labor of merging people again.

**Comment by rockinyp** [comment]
I'd like to vote for this. I have identical twin girls and they're identified as the same person in all of their photos. I can adjust the ML settings and get them to be two separate people, but then I end up with a LOT of copies of everyone else that I have to continually merge. It would be easier if I could just split my twins manually.

**Comment by triantos** [comment]
+1. My twins are also merged together.
